prover/src/constraints/evaluator/default.rs

```diff
@@ -257,6 +288,41 @@ where
         }
     }
 
+    fn evaluate_lagrange_kernel_constraints<T: TraceLde<E>>(
+        &self,
+        trace: &T,
+        domain: &StarkDomain<A::BaseField>,
+    ) -> Vec<E> {
+        let lagrange_kernel_aux_column_idx = self
+            .air
+            .context()
+            .lagrange_kernel_aux_column_idx()
+            .expect("expected Lagrange kernel aux column index to be present");
+
+        // this will be used to convert steps in constraint evaluation domain to steps in
+        // LDE domain
+        let lde_shift = domain.ce_to_lde_blowup().trailing_zeros();
+
+        let mut lagrange_kernel_column_frames =
+            vec![LagrangeKernelEvaluationFrame::<E>::new_empty(); domain.ce_domain_size()];
+
+        for (step, frame) in lagrange_kernel_column_frames.iter_mut().enumerate() {
+            trace.read_lagrange_kernel_frame_into(
+                step << lde_shift,
+                lagrange_kernel_aux_column_idx,
+                frame,
+            );
+        }
```

Comment:
If I'm understanding this correctly, this would store a frame per row in the constraint evaluation domain, right? If so, this is probably too expensive from memory standpoint. For example, for a trace with $2^20$ columns, this would be equivalent to having extra 20 columns in the auxiliary trace, and this would significantly increase RAM requirements (e.g., in some cases doubling them).

This interferes a bit with batch inversion of denominators approach (would have been really nice if we could pre-compute them for all constraints in a single vector) - but I think we can still make it work if we change the methodology slightly: we still evaluate one constraint at a time, but instead of working with evaluation frames, we work directly with the trace (e.g., read elements needed only for the current constraint). Maybe there are other ways of doing this too.